src/crimson/osd/object_context_loader.cc

```diff
@@ -146,7 +146,7 @@ using crimson::common::local_conf;
     LOG_PREFIX(ObjectContextLoader::get_or_load_obc);
     auto loaded =
       load_obc_iertr::make_ready_future<ObjectContextRef>(obc);
-    if (existed) {
+    if (existed && obc->is_loaded()) {
```

Comment:
I think this is the correct direction to return only until the `obc` is loaded.

> As obc-loading is now concurrent, it's possible for the first user to only create the obc in cache (without loading yet) and the second concurrent user to assume it was already loaded.

But I think the first user must create and also start to load the `obc` at the same time. So the second (maybe-concurrent) user should assume it is already loading or already loaded.

(If I'm wrong above, can you help point to the violating logic?)

If I'm correct above, I think the accurate pseudo logic might be:
```c++
if (existed && obc->is_loaded()) {
  return load_obc_iertr::make_ready_future<ObjectContextRef>(obc);
} else if (existed) {
  // not the first user, the obc must be already loading,
  // return the loading future, instead of loading the obc again.
} else { // !existed: the first user
  return obc->template with_promoted_lock<State, IOInterruptCondition>(
    [obc, this] { return load_obc(obc); }
  );
}
```

However, the current version should be fine to address the assert failure.

Reply:
>But I think the first user must create and also start to load the obc at the same time. So the second (maybe-concurrent) user should assume it is already loading or already loaded.

This is correct. Broadly speaking, the pseudo logic should work, though I'm now sure it's worth the additional logic. This will require maintaining the loading future from the "first-user" for the "second" one to use later, instead of loading again.
The main reason for avoiding implementing the pseudo logic is that this scenario is rare. This was quite frequent in testing was because the obc_lru_size is set to 10 (to reduce cache hits).

@cyx1231st, I suggest merging as is to resolve the bug. Later on, we could rework this bit if needed. What do you think?

Reply:
> This is correct. Broadly speaking, the pseudo logic should work, though I'm now sure it's worth the additional logic. This will require maintaining the loading future from the "first-user" for the "second" one to use later, instead of loading again.


@Matan-B https://github.com/ceph/ceph/pull/57687 is for this purpose, it avoid loading the obc again by making sure the `exist` check and the load happens synchronously, maybe we should consider it?

Reply:
> @Matan-B #57687 is for this purpose, it avoid loading the obc again by making sure the `exist` check and the load happens synchronously, maybe we should consider it?

@xxhdx1985126 , To consider the benefit of avoiding the extra load we need to understand how frequent this scenario is. IIUC, a normal workload (not thrash test), with a decent size of obc_lru_size (default of 512) - should not be impacted by this.
For this reason, I'm leaning towards readability and maintainability here.

Reply:
Note: I'm scheduling a test run since this assert causes too many jobs to fail and merging is blocked, please let me know if you don't agree with my previous comments.

Reply:
@Matan-B @cyx1231st I think this pr may fail. According to the following code, we can't promote the lock to EXCL when there are multiple readers/writers. Am I missing something?

```
void tri_mutex::promote_from_read()    
{                                      
  assert(readers == 1);                
  --readers;                           
  exclusively_used = true;             
}                                      

void tri_mutex::promote_from_write()    
{                                       
  assert(writers == 1);                 
  --writers;                            
  exclusively_used = true;              
}                                       
```

Reply:
> @cyx1231st, I suggest merging as is to resolve the bug. Later on, we could rework this bit if needed. What do you think?

I'm fine to address the specific assert failure with a simple fix, and then find a proper way to avoid loading the same obc.

And I think it is also related to https://tracker.ceph.com/issues/65451, because allowing the non-first user(s) to load the same obc (in the promoted/exclusive lock) should violate the assertion `readers/writers == 1' upon promotion.

> This will require maintaining the loading future from the "first-user" for the "second" one to use later, instead of loading again. The main reason for avoiding implementing the pseudo logic is that this scenario is rare. ..

A per-obc `seastar::shared_mutex<>` may be helpful to make the concurrent users (except the first user) to wait outside the promoted/exclusive lock until the obc is loaded, and also preserve the waiting order. Otherwise, we might need to rethink tri-mutex and the lock promotion. Anyway, the exclusive constraint to load the same obc makes more sense to me.